### PR TITLE
Provisioner for CoreOS

### PR DESCRIPTION
**Pull request by mschygulla**
Added support for CoreOS

### CODE REVIEW

**Comment by UserTaken** [comment]
Seems to be missing `$DOCKER_OPT_BIP`, `$DOCKER_OPT_MTU`, and `$DOCKER_OPT_IPMASQ` variables to use Flannel config in `/run/flannel_docker_opts.env`.

File: libmachine/provision/coreos.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+		EngineOptions:     engineCfg.String(),
+		EngineOptionsPath: provisioner.DaemonOptionsFile,
+	}, nil
+}
+
+func (provisioner *CoreOSProvisioner) Package(name string, action pkgaction.PackageAction) error {
+	return nil
+}
+
+func (provisioner *CoreOSProvisioner) dockerDaemonResponding() bool {
```
Comment (nathanleclaire) [inline libmachine/provision/coreos.go]: Any reason that `utils.WaitForDocker` won't do the job for you here?

**Comment by nathanleclaire** [comment]
Cool - how have you been testing this?  Using the image(s) on DigitalOcean?

File: libmachine/provision/coreos.go (hunk 1 above)
Comment (mschygulla) [inline libmachine/provision/coreos.go]: Actually I though this is a safe way to test if docker is up and running and thats why other provisioner are doing it. But I've tested `utils.WaitForDocker` and it seem to work fine. I will update my code.

**Comment by mschygulla** [comment]
So far I've tested the provisioner locally with with an improved VMware Fusion driver see PR https://github.com/docker/machine/pull/1479

**Comment by dudymas** [comment]
Confirmed this works on openstack kilo with coreos 675.0.0
Thanks for this!

**Comment by dudymas** [comment]
Okay, so I ran 'update_engine_client -update' and forced coreos to upgrade to 675.0.0

CoreOS then invalidated its TLS keys. I guess I wanted to mention this in case anyone else gets an authentication error, you'll need to run `docker-machine regenerate-certs $MACHINE_NAME` if you want to fix that.

**Comment by JeanMertz** [comment]
:+1: for this

**Comment by FabienMotte** [comment]
Any news about this PR ? (:

**Comment by nathanleclaire** [comment]
We're interested, just have been busy with the release and not able to do a thorough review.

**Comment by FabienMotte** [comment]
@nathanleclaire I totally understand, thanks for your answer anyway !

**Comment by ehazlett** [comment]
This looks pretty good to me.  

Would you mind moving the `SetHostname` output to debug?

Tested on DigitalOcean (note: if you want to test you will need this https://github.com/docker/machine/pull/1667 to specify the ssh user):

```
ejh-x1:ehazlett> docker-machine create -d digitalocean --digitalocean-image coreos-alpha --digitalocean-ssh-user core coreos

Creating SSH key...
Creating Digital Ocean droplet...
SetHostname: coreos
To see how to connect Docker to this machine, run: docker-machine_linux-amd64 env coreos

ejh-x1:ehazlett machine> docker-machine ls
NAME              ACTIVE   DRIVER         STATE     URL                       SWARM
coreos                     digitalocean   Running   tcp://45.55.70.227:2376

ejh-x1:ehazlett machine> docker $(docker-machine config coreos) info
Containers: 0
Images: 0
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.0.9-coreos
Operating System: CoreOS 758.1.0
CPUs: 1
Total Memory: 997.6 MiB
Name: coreos
ID: MTPZ:TS2L:YBRP:6T2T:WV2G:KB4X:YLGB:ZJXC:7B45:HEWX:2BVH:ZVUS
Username: ehazlett
Registry: https://index.docker.io/v1/
Labels:
 provider=digitalocean
```

**Comment by mschygulla** [comment]
Sorry, my mistake. Log output fixed.

**Comment by ehazlett** [comment]
@mschygulla thanks!  would you mind squashing the commits down to a single commit?  i think after that we should be good.  thanks!

**Comment by mschygulla** [comment]
Yep, done.

**Comment by ehazlett** [comment]
LGTM thanks!

**Comment by sylus** [comment]
Theoretically should this work for Amazon EC2?

**Comment by sylus** [comment]
I switched to digitalocean and this worked to boot up coreos. However without etcd configured I am not sure about the utility of this provisioner? Am I correct in that this is just the first step and more work needs to be done with cloud-init so discovery token would configure etcd.

**Comment by mschygulla** [comment]
@sylus, sorry haven't seen your question earlier.
The coreos provisioner configures only what's needed to use the docker service remotely. If you want to configure coreos or other services like the etcd, then the best way is to use the cloud-init. For AWS you find some instructions here: https://coreos.com/os/docs/latest/booting-on-ec2.html

**Pull request merged**